### ISSUE TITLE
Elimination methods in classes should be moved

### ISSUE THREAD

**Issue by Ericson2314** [tweag/capability#63]
Methods like `local`, `catch`, `listen`, etc, are awkward holdovers from mtl. They are effect *handlers*, in that they operationaly eliminate the main effect of the class (reading, throwing, writing, etc.). This is alright in mtl's intensional model (to use the blog post's term), but out of place in capability's extensional model.

The downsides see practical, not just philosophical. Firstly, It's weird that there is no way to hand out the capability without also handing out the addition capability to intercept the former and implement it another way. There is no technical reason to be so coarse grained. Intuitively, handling an effect is a more privileged operation: I can raise a complaint to the cops or HR, but only they can dismiss it (throw, catch). I can forward packets and mail but I shouldn't modify them (ask, local). I can flush the toilet in my apartment but cannot receive my neighbors sewage (tell, listen). MTL makes me citizen-cop, mailman-censor, tenent-plumber, but never just the first of each pair.

Secondarily, it's unfortunate that the types hide that the effect is in fact eliminated (a catch application needs the original `HasThrow`, a `local` application needs the original `HasReader`, etc.). However things may happen operationaly, at the type level effects/constraints are only accumulated, until the final grand discharge in `main`.

In an ideal world, I'd rip these methods out and indeed replace them with some other mechanism to prove the elimination of the effect, like running the underlying transformer. I don't see how to do that in all cases though without incurring other downsides, so instead I advocate the simpler approch of splitting the classes: Just as `MonadError` is spilit into `MonadThrow` and `MonadCatch`, so split the others to separate the introduction of each effect and it's elimination. At least this solves the first problem, and isolates the second. If/when a better way of handling effects comes along, code that only introduces effects (throws/asks/etc.) can stay the same.

**Comment by Ericson2314** [comment]
Right I think we knowing agree acting in my second issue is premature. But I think everything you said is compatible with taking action on the first point as I allow.

> I'd like to point out that despite the similarities, capability does not try to follow the precepts of algebraic effects

Sure. I'm not deeply familiar with that stuff anyways. I suppose "handler" is their term, but I am using it imprecisely.

> I'm also ok with splitting capabilities further if there is a need for it.

Great!

> The reason why HasThrow and HasCatch have been separated in two class is that it's common for a monad to be provide throw without being able to provide catch.

Note that the fact that anyone bothered to create those monads in code at all implies it is also commonly useful to be able to throw without catching, or commonly useful enforce that other code throws without catching. I'm arguing that in the case of the other classes, even if no concrete monad implemented only half the class, it is still valid to restrict code to half the class.

> But I think that even if we decide to have comprehensive support for handlers, we may still want to also keep listen, catch and local as effects: they're useful!

Sure! And that's another reason why let's keep `HasCatch` and the others I propose creating, as opposed to just thinking of them as stop-gaps. 

> If there are such examples for `Reader` and `Writer` (well, `HasWriter` without a listen is kind of the same as `HasStream`, I suppose, so there is an example here) then we should consider it.

Maybe making `HasWriter` a subclass of `HasStream` like `HasCatch` is of `HasThrow` is a good place to start?

**Comment by aspiwack** [comment]
>  I'm arguing that in the case of the other classes that it is still valid to restrict code to half the class, even if no concrete monad implements only half the class.

I see where you come from now. I'd still like this to be a bit better supported. Typically with examples. There is a cost to splitting type classes down. So the benefit had better not be hypothetical.

To be clear: I have no particular attachment to the way capability is currently arranged, and I would be happy to change anything in the API if it proves better. As long as it's justified, it's good with me.

> Maybe making HasWriter a subclass of HasStream

Maybe. Though there is an interesting point to note, which is that `HasStream` has no law, but in `HasWriter`, we at least have compatibility with the monoid operations  (`put mempty = return ()` and `put a >> put b = put (a<>b)`). It may not be an obstacle. But it shows that in reality, `HasStream` is not quite the same thing as `HasWriter` without listen.

**Comment by Ericson2314** [comment]
Awesome yeah I think we are on the same page. I have no killer app for split reader; the use-case I believe must always isomorphic to wanting to ensure everyone `asks` the same exact thing. So yes, let's ignore than one for now and just worry about stream and writer.

> Which is that `HasStream` has no law, but in `HasWriter`, we at least have compatibility with the monoid operations

Firstly, I have no problem with subclasses adding laws only involving superclasses' methods. Anecdotally, this seems to come up a lot. And if one imagines new libraries with dependent Haskell, the law proof can exist in the subdictionary just fine either way.

That said, I think it's interesting to compare what you wrote to how it's phrased in the source https://github.com/tweag/capability/blob/master/src/Capability/Writer.hs#L69 . In particular, to *witness* this equality or it's violation, we either need to `runWriterT` or use `listen`. Granted laws are important whether or not they directly observable, but I think there's something to be said here that a pure `tell` class ought not to concern itself with `Monoids` as `HasStream` does, so the "intemediate" step of `(HashStream tag w a, MonoidConstraintWithLaws tag w a)` isn't worth worrying about or considering a separator between `HasStream` and `HasWriter`.

Put simply, I see no moral conflict between
````haskell
forall a. HasStream a => HasWriter [a] -- free monoid, super class not relevant
exists a. HasStream a => HasWriter a
```

**Comment by Ericson2314** [comment]
All my fluff aside, another simple practical argument is if I start with `HasStream`, and then decide I need to go to `HasWriter`, I don't want to have to bother refactoring all my `yield`s to `tell`s, and `HasStream` constraints to `HasWriter` constraints. This makes the migration smooth, and preserves the evidence to the degree to which `HasWriter` is needed.

**Comment by Ericson2314** [comment]
Implemented the `HasStream` => `HasWriter` part in https://github.com/tweag/capability/pull/64 .

**Comment by Ericson2314** [comment]
What about Conduit's `await` (http://hackage.haskell.org/package/conduit-1.3.1/docs/Data-Conduit.html#t:Consumer)? There is a `leftover` dual which puts back, but that is not a candidate for  `local` or State `put`. And similar imperative/foreign-interfacing stream consumers do not have such a thing. If Anything we could have have:
```haskell
class HasStreamOut tag w m => HasWriter tag w m
class HasStreamIn tag r m => HasReader tag r m
-- no methods, but LIFO laws
class (HasStreamIn tag c m, HasStreamOut tag c m) => HasLifo tag c m
```
This gives us
```haskell
instance HasFifo "input" i => ConduitT i o m
instance HasStreamOut "output" o => ConduitT i o m
```

**Comment by Ericson2314** [comment]
I finally found time at Compose NYC to get back to this. https://github.com/tweag/capability/pull/75 is the PR.
```
    Sink     Source
    /   \    /   \
   /     \  /     \
Writer   State    Reader
```
Is the new class hierarchy (`Sink` is rename of `Stream`). `Conduit`, per my old example from December, can implement `Source`, but not `Reader`.
